intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/verifiers/PluginUntilBuildVerifier.kt

```diff
@@ -0,0 +1,123 @@
+package com.jetbrains.plugin.structure.intellij.verifiers
+
+import com.jetbrains.plugin.structure.intellij.beans.PluginBean
+import com.jetbrains.plugin.structure.intellij.problems.InvalidUntilBuild
+import com.jetbrains.plugin.structure.intellij.problems.InvalidUntilBuildWithJustBranch
+import com.jetbrains.plugin.structure.intellij.problems.InvalidUntilBuildWithMagicNumber
+import com.jetbrains.plugin.structure.intellij.problems.NonexistentReleaseInUntilBuild
+import com.jetbrains.plugin.structure.intellij.problems.ProductCodePrefixInBuild
+import com.jetbrains.plugin.structure.intellij.problems.SuspiciousUntilBuild
+import com.jetbrains.plugin.structure.intellij.verifiers.PluginUntilBuildVerifier.ValidationResult.INVALID
+import com.jetbrains.plugin.structure.intellij.verifiers.PluginUntilBuildVerifier.ValidationResult.VALID
+import com.jetbrains.plugin.structure.intellij.version.IdeVersion
+
+
+private const val BUILD_NUMBER = "__BUILD_NUMBER__"
+private const val SNAPSHOT = "SNAPSHOT"
+
+private const val SUSPICIOUS_BASELINE_LOWER_BOUND = 281
+private const val FIRST_YEARLY_BASED_RELEASE_NUMBER_BASELINE = 162
+private const val FIRST_YEARLY_BASED_RELEASE_NUMBER_YEAR = 2016
+
+class PluginUntilBuildVerifier {
+  fun verify(plugin: PluginBean,
+             descriptorPath: String,
+             problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    val untilBuild = plugin.ideaVersion?.untilBuild ?: return
+    if (isJustASingleComponent(untilBuild)) {
+      if (isSpecialSingleComponent(untilBuild)) {
+        return
+      }
+      if (verifyMagicNumber(untilBuild, descriptorPath, problemRegistrar) == INVALID) {
+        return
+      }
+      registerProblem(InvalidUntilBuildWithJustBranch(descriptorPath, untilBuild))
```

Comment:
It seems that `InvalidUntilBuildWithJustBranch` was covered by `IdeVersion.createIdeVersionIfValid(untilBuild)`. Do we really need this as a separate problem?

Reply:
I mean, maybe `isJustASingleComponent` is not needed at all

Reply:
Rationale: plugin developers are somewhat confused by the semantics. Especially with different behaviour than happens in`since-build`. Hence the special problem and a special explanation in the message.

- `until-build` prohibits `223`. It should be `223.*`. (`InvalidUntilBuildWithJustBranch`)
- `since-build` prohibits `223`. It should be `223.*`.

Reply:
In addition, `createIdeVersionIfValid` has a very barebones handling. It just returns `null` when the version cannot be parsed. This tries to be more granular and more helpful.

Reply:
Since like `223.*` is not correct also. Probably you mean something else. But I see the point. 👍

If I set until as `999` now I get two problems: `InvalidUntilBuildWithJustBranch` and `InvalidUntilBuildWithMagicNumber`, is it ok?

Reply:
`999` triggers an `InvalidUntilBuildWithMagicNumber`.

There is a dedicated [unit test](https://github.com/JetBrains/intellij-plugin-verifier/blob/a70f231065b920c17d6bef4abbdcfecf8bd27617/intellij-plugin-structure/tests/src/test/kotlin/com/jetbrains/plugin/structure/mocks/InvalidPluginsTest.kt#L670) for that.

Fixed the original _Rationale_, there was a confusion even on my side.

Reply:
Aaah, `verifyMagicNumber` 🤯